### PR TITLE
Port NetworkProtocolCurl to new network2::Network interface

### PR DESCRIPTION
**Pull request by serhiikobalt**
Signed-off-by: Serhii Lysenko <ext-serhii.lysenko@here.com>

### CODE REVIEW

File: olp-cpp-sdk-core/include/olp/core/network2/Network.h [hunk 1]
```diff
@@ -67,12 +69,17 @@ class CORE_API Network {
                          HeaderCallback header_callback = nullptr,
                          DataCallback data_callback = nullptr) = 0;
 
+  NetworkResponse SendAndWait(NetworkRequest request,
```
[the remark this answers is not in the document]
Comment (sergii-vostrikov) [reply olp-cpp-sdk-core/include/olp/core/network2/Network.h]: Network2 Design draft document
Requirements
*Network stack functional requirements*
- Synchronous & Asynchronous API

File: olp-cpp-sdk-core/include/olp/core/network2/NetworkTypes.h [hunk 2]
```diff
@@ -35,14 +35,12 @@ using RequestId = std::uint64_t;
...
  */
-enum class RequestIdConstants : std::uint64_t {
-  /// Value that indicates invalid request id.
-  RequestIdInvalid = std::numeric_limits<RequestId>::min(),
-  /// Minimal value of valid request id.
-  RequestIdMin = std::numeric_limits<RequestId>::min() + 1,
-  /// Maximal value of valid request id.
-  RequestIdMax = std::numeric_limits<RequestId>::max(),
-};
+namespace RequestIdConstants {
```
[the remark this answers is not in the document]
Comment (sergii-vostrikov) [reply olp-cpp-sdk-core/include/olp/core/network2/NetworkTypes.h]: enum class has one big (and the only) advantage - it disallows transparent "int" operations.
Unfortunately, this is exactly what's required for RequestId counter.

File: olp-cpp-sdk-core/src/network2/NetworkCurl.cpp [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+  }
+  char unused;
+  while (read(worker_notification_pipe_[0], &unused, sizeof(unused)) > 0)
+    ;
+  return;
+}
+
+size_t curl_write_callback(char* ptr, size_t size, size_t nmemb,
+                           void* userdata) {
+  RequestContext& context = *reinterpret_cast<RequestContext*>(userdata);
```
Comment (sergii-vostrikov) [inline olp-cpp-sdk-core/src/network2/NetworkCurl.cpp]: assert userdata for non-zero?

File: olp-cpp-sdk-core/src/network2/NetworkCurl.cpp [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+      curl_easy_reset(easy_handle);
+      running_requests_[easy_handle]->callback(
+          NetworkResponse().WithCancelled(true));
+      running_requests_.erase(easy_handle);
+      free_handles_.push_back(easy_handle);
+      continue;
+    }
+
+    if (request_queue_.find(request_id) != request_queue_.end()) {
+      request_queue_[request_id]->callback(
```
Comment (sergii-vostrikov) [inline olp-cpp-sdk-core/src/network2/NetworkCurl.cpp]: find's iterator can be utilized here and at erasure

File: olp-cpp-sdk-core/src/network2/NetworkCurl.cpp [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+                            Callback callback, HeaderCallback header_callback,
+                            DataCallback data_callback) {
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  if (next_request_id_ == RequestIdConstants::RequestIdMax) {
+    next_request_id_ = RequestIdConstants::RequestIdMin;
+  } else {
+    next_request_id_++;
+  }
+  std::unique_ptr<RequestContext> request_context(
```
Comment (sergii-vostrikov) [inline olp-cpp-sdk-core/src/network2/NetworkCurl.cpp]: std::make_unique

File: olp-cpp-sdk-core/src/network2/NetworkCurl.cpp (hunk 3 above)
Comment (andescu) [reply olp-cpp-sdk-core/src/network2/NetworkCurl.cpp]: If only then nullptr checks and log/return.
Aborting/assert on edge devices is a serious sin and almost always disabled.

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 6]
```diff
@@ -0,0 +1,1297 @@
...
+    default:
+      assert(false);
+      return CURLPROXY_HTTP;
+  }
+}
+}  // namespace
+
+namespace olp {
+namespace network2 {
+#define LOGTAG "CURL"
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]:
```suggestion
const char * kLogTag = "CURL";
```

File: olp-cpp-sdk-core/src/network2/Memory.h [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+struct MemoryScopeTracker {
+  MemoryScopeTracker() = default;
+  explicit MemoryScopeTracker(bool) {}
+  void Capture() {}
+  void Clear() {}
+};
+
+}  // namespace mem
+}  // namespace network
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/Memory.h]:
```suggestion
}  // namespace network2
```

File: olp-cpp-sdk-core/src/network2/Memory.h [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ * @brief Memory tracking definitions.
+ *
+ * If NETWORK_USE_MEMORY_LIB is non-zero then the memory library tracking
+ * mechanism will be used.
+ *
+ * Otherwise the relevant macros, classes and type-aliases will be redefined to
+ * have no effect thus eliminating memory library dependencies.
+ */
+
+#include <deque>
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/Memory.h]: Do we need those includes here?

File: olp-cpp-sdk-core/src/network2/Memory.h [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  explicit MemoryScopeTracker(bool) {}
+  void Capture() {}
+  void Clear() {}
+};
+
+}  // namespace mem
+}  // namespace network
+}  // namespace olp
+
+#define MEMORY_TRACKER_SCOPE(tracker)
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/Memory.h]:
```suggestion
#define EDGE_SDK_MEMORY_TRACKER_SCOPE(tracker)
```

File: olp-cpp-sdk-core/src/network2/Memory.h [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+#define MEMORY_SCOPED_TAG(name)
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/Memory.h]:
```suggestion
#define EDGE_SDK_MEMORY_SCOPED_TAG(name)
```

File: olp-cpp-sdk-core/src/network2/Memory.h [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+#define MEMORY_GET_TAG_ALLOCATOR(name, type) \
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/Memory.h]:
```suggestion
#define EDGE_SDK_MEMORY_GET_TAG_ALLOCATOR(name, type) \
```

File: olp-cpp-sdk-core/src/network2/NetworkUtils.cpp [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  while (offset < str1.length() - str2.length()) {
+    if (SimpleToUpper(str1[offset]) == SimpleToUpper(str2[0])) {
+      if (CaseInsensitiveStartsWith(str1, str2, offset)) return offset;
+    }
+    offset++;
+  }
+  return std::string::npos;
+}
+
+}  // namespace network
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/NetworkUtils.cpp]:
```suggestion
}  // namespace network2
```

And in other files as well

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 13]
```diff
@@ -0,0 +1,1297 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#include "NetworkProtocolCurl.h"
+
+#include <algorithm>
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]: Some of these includes are already included in `NetworkProtocolCurl.h`. Please, double check them and remove excessive ones, like:
```
#include <memory>
#include <mutex>
#include <string>
#include <thread>

#include <curl/curl.h>
```

File: olp-cpp-sdk-core/src/network2/NetworkUtils.h [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+                                     size_t offset = 0);
+  static bool CaseInsensitiveStartsWith(const std::string& str1,
+                                        const std::string& str2,
+                                        size_t offset = 0);
+  static size_t CaseInsensitiveFind(const std::string& str1,
+                                    const std::string& str2, size_t offset = 0);
+
+};  // class NetworkUtils
+
+}  // namespace network
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/NetworkUtils.h]:
```suggestion
}  // namespace network2
```

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 15]
```diff
@@ -0,0 +1,1297 @@
...
+
+  if (!IsStarted()) {
+    return ErrorCode::OFFLINE_ERROR; //NetworkProtocol::ErrorNotReady;
+  }
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    AddEvent(EventInfo::Type::SendEvent, handle);
+  }
+  return ErrorCode::NO_ERROR; //NetworkProtocol::ErrorNone;
+}  // namespace network
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]:
```suggestion
}
```
It just a closing bracket for Send2 method

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.h [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+  /**
+   * @brief Free resources after the thread terminates.
+   */
+  void Teardown();
+
+  void AddEvent(EventInfo::Type type, RequestHandle* handle);
+
+  RequestHandle handles_[kTotalHandleCount] = { };
+  std::condition_variable event_condition_;
+  std::mutex mutex_;
```
Comment (Dschoordsch) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.h]: A more descriptive name like `event_mutex_` would be nice, currently I have to assume what it's used for.

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 17]
```diff
@@ -0,0 +1,1230 @@
...
+  close(pipe_[1]);
+#endif
+
+  // Handle completed messages
+  if (!completedMessages.empty()) {
+    for (auto& pair : completedMessages) {
+      pair.second(network2::NetworkResponse()
+        .WithRequestId(pair.first)
+        .WithStatus(static_cast<int>(ErrorCode::OFFLINE_ERROR))
+        .WithError("Offline"));
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]:
```suggestion
        .WithError("Offline: network is deintialized"));
```
I guess it is better to extend the error string that the error happened during teardown (and to distinguish this case from regular Offline).

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 18]
```diff
@@ -0,0 +1,1230 @@
...
+  close(pipe_[0]);
+  close(pipe_[1]);
+#endif
+
+  // Handle completed messages
+  if (!completedMessages.empty()) {
+    for (auto& pair : completedMessages) {
+      pair.second(network2::NetworkResponse()
+        .WithRequestId(pair.first)
+        .WithStatus(static_cast<int>(ErrorCode::OFFLINE_ERROR))
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]:
```suggestion
        .WithStatus(static_cast<int>(ErrorCode::CANCELLED_ERROR))
```
Maybe cancelled error is better for such scenario?

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 19]
```diff
@@ -0,0 +1,1230 @@
...
+
+  if (error_status == ErrorCode::NO_ERROR) {
+    return NetworkStatus(request_id);
+  }
+
+  return NetworkStatus(error_status);
+}
+
+
+ErrorCode NetworkProtocolCurl::Send2(
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]: My personal opinion is to not include adapter between Network2 new interface and old implementation in the scope of this task and do it in separate commit / pull request. 

Thus, commits will be cleaner and you can easier to track how original Network2 interface looked like according to our design document.

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 20]
```diff
@@ -0,0 +1,1230 @@
...
+    const std::shared_ptr<std::ostream>& payload,
+    Network::HeaderCallback headerCallback, Network::DataCallback dataCallback,
+    Network::Callback callback) {
+  EDGE_SDK_LOG_TRACE(kLogTag, "send with id = " << id);
+
+  if (!IsStarted()) {
+    return ErrorCode::IO_ERROR;
+  }
+
+    const auto& config = request.GetSettings();
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]:
```suggestion
  const auto& config = request.GetSettings();
```
Run clang-format on this file.

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 21]
```diff
@@ -0,0 +1,1230 @@
...
+  }
+}
+}  // namespace
+
+namespace olp {
+namespace network2 {
+const char* kLogTag = "CURL";
+
+namespace {
+int convertErrorCode(CURLcode curlCode) {
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]:
```suggestion
int ConvertErrorCode(CURLcode curl_code) {
```
Also, make sure that all code aligned with Google C++ code style

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 22]
```diff
@@ -0,0 +1,1230 @@
...
+      gSslMutexes[n].lock();
+    } else {
+      gSslMutexes[n].unlock();
+    }
+  }
+}
+
+unsigned long sslIdFunction(void) {
+  std::hash<std::thread::id> hasher;
+  return ((unsigned long)hasher(std::this_thread::get_id()));
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]:
```suggestion
  return static_cast<unsigned long>(hasher(std::this_thread::get_id()));
```

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 23]
```diff
@@ -0,0 +1,1230 @@
...
+  handle->in_use = false;
+  handle->callback = nullptr;
+  handle->header_callback = nullptr;
+  handle->data_callback = nullptr;
+  handle->payload.reset();
+}
+
+size_t NetworkProtocolCurl::RxFunction(void* ptr, size_t size, size_t nmemb,
+                                       RequestHandle* handle) {
+
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]: Remove unneeded spaces

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 24]
```diff
@@ -0,0 +1,1230 @@
...
+
+
+
+  size_t len = size * nmemb;
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]:
```suggestion
  const size_t len = size * nmemb;
```

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 25]
```diff
@@ -0,0 +1,1230 @@
...
+    }
+  }
+
+  return len;
+}
+
+size_t NetworkProtocolCurl::HeaderFunction(char* ptr, size_t size,
+                                           size_t nitems,
+                                           RequestHandle* handle) {
+
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]: Remove unneeded spaces

File: olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp [hunk 26]
```diff
@@ -0,0 +1,1230 @@
...
+
+
+  size_t len = size * nitems;
```
Comment (borzun) [inline olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.cpp]:
```suggestion
  const size_t len = size * nitems;
```

File: olp-cpp-sdk-core/include/olp/core/network2/Network.h [hunk 27]
```diff
@@ -79,5 +79,7 @@ class CORE_API Network {
   virtual void Cancel(RequestId id) = 0;
 };
 
+std::shared_ptr<Network> CreateNetwork();
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/network2/Network.h]: remove. this will be implemented in OlpClientSettingsFactory, see #80

File: olp-cpp-sdk-core/cmake/curl.cmake [hunk 28]
```diff
@@ -18,6 +18,8 @@
 if(CURL_FOUND AND NOT NETWORK_NO_CURL)
     aux_source_directory(${CMAKE_CURRENT_LIST_DIR}/../src/network/curl NETWORK_CURL_SOURCES)
     set(NETWORK_CURL_SOURCES ${NETWORK_CURL_SOURCES} ${CMAKE_CURRENT_LIST_DIR}/../src/network/socket/NetworkConnectivitySocketImpl.cpp)
+    set(NETWORK2_CURL_SOURCES ${NETWORK2_CURL_SOURCES} ${CMAKE_CURRENT_LIST_DIR}/../src/network2/curl/NetworkCurl.cpp)
```
Comment (andescu) [inline olp-cpp-sdk-core/cmake/curl.cmake]: Variable name changed to EDGE_SDK_HTTP_CURL_SOURCES, please adapt.

File: olp-cpp-sdk-core/include/olp/core/utils/Dir.h [hunk 29]
```diff
@@ -55,6 +55,12 @@ class CORE_API Dir {
    * @return temporary directory path
    */
   static std::string TempDirectory();
+
+  /**
+   * @brief Check whether file exists.
+   * @return @c true if file with a given name exists, @c false otherwise.
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/utils/Dir.h]:
```suggestion
   * @return \c true if any file with the given path exists, \c false otherwise.
```

File: olp-cpp-sdk-core/include/olp/core/utils/Dir.h [hunk 30]
```diff
@@ -55,6 +55,12 @@ class CORE_API Dir {
...
+   */
+  static bool FileExists(const std::string& file_path);
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/utils/Dir.h]: add also doxygen for file_path
